ott/geometry/costs.py

```diff
@@ -20,11 +20,83 @@
 
 import jax
 import jax.numpy as jnp
+from numpy import int16
 
 from ott.core import fixed_point_loop
 from ott.geometry import matrix_square_root
 
 
+def x_to_means_and_covs(x, dimension):
+  """Extract means and covariance matrices of Gaussians from raveled vector.
+
+  Args:
+    x: [num_gaussians, dimension, (1 + dimension)] jnp.ndarray of concatenated means and covariances (raveled)
+    dimension: the dimension of the Gaussians
+  Returns:
+    means: [num_gaussians, dimension] jnp.ndarray that holds the means.
+    covariances: [num_gaussians, dimension] jnp.ndarray that holds the covariances.
+  """
+  x = jnp.atleast_2d(x)
+  means = x[:, 0:dimension]
+  covariances = jnp.reshape(
+      x[:, dimension:dimension + dimension ** 2], (-1, dimension, dimension)
+  )
+  return jnp.squeeze(means), jnp.squeeze(covariances)
+
+
+def mean_and_cov_to_x(mean, covariance, dimension):
+  """Ravel a Gaussian's mean and covariance matrix to d(1 + d) vector."""
+  x = jnp.concatenate((mean, jnp.reshape(covariance, (dimension * dimension))))
+  return x
+
+
+means_and_covs_to_x = jax.vmap(mean_and_cov_to_x, in_axes=[0, 0, None])
+
+
+def add_identity(A):
+  return A + jnp.eye(A.shape[0])
+
+
+add_identities = jax.vmap(add_identity, in_axes=0)
+
+
+# TODO(ersi):adapt to use also for initialization of barycenter with Bures cost
+def mean_and_cov_segmenter(
+    x, a, segment_ids, num_segments, num_per_segment, max_measure_size
+):
+  """Segmenter that pads with concatenated zero means and raveled identity covariance matrices."""
+  num, dim = x.shape
+  # obtain the dimension of the Gaussians from the dimension of the pointcloud.
+  dimension = jnp.array((-1 + jnp.sqrt(1 + 4 * dim)) / 2, dtype=int16)
+  segmented_a = []
+  segmented_x = []
+  a = jnp.concatenate((a, jnp.zeros((1,))))
+  for i in range(num_segments):
+    # segment the weights a
+    idx = jnp.where(segment_ids == i, jnp.arange(num), num + 1)
+    idx = jax.lax.dynamic_slice(jnp.sort(idx), (0,), (max_measure_size,))
+    z = a.at[idx].get()
+    segmented_a.append(z)
+
+    # segment the positions x
+    idx = jnp.where(segment_ids == i)[0]
+    z = x.at[idx, :].get()
+    padding = means_and_covs_to_x(
+        jnp.zeros((max_measure_size - num_per_segment[i], dimension)),
+        add_identities(
+            jnp.zeros(
+                (max_measure_size - num_per_segment[i], dimension, dimension)
+            )
+        ), dimension
+    )
+    z = jnp.vstack((z, padding))
+    segmented_x.append(z)
```

Comment:
i think the custom segmented has a lot of copy pasted code from the regular one. It seems to me the main particularity of this novel segmentation approach is to replace simple zero padding with slightly more clever padding. couldn’t we think of only passing a “zero” to the segmenter, rather than take for granted it is jnp.zeros ? here the goal is to minimize any duplicated code

Reply:
The segmentation for the weights is exactly the same as the regular one, which makes sense since the padded points should always have mass equal to zero. So, that won't change. The segmentation of the positions of the point cloud (means and covariances here) is different. An option would be to only do the custom segmentation for the point cloud, The only reason why I avoided this is because I would then need to have two for loops in the regular segmented, which would share a lot of code with another. WDYT? I am not sure I understand the comment regarding passing a zero to the segmenter.

Reply:
modified to custom_padder instead of custom_segmenter in order to reuse code better. I still think that in the future we might want to re-organize padders and initializers in order to re-use code  in a tidier way (see my comment above). Happy to talk more about this. Let me know if you prefer this solution for now.